### PR TITLE
prov/tcp: add support for getpeer() function

### PR DESCRIPTION
**Pull request by sydidelot**
This patch adds the support of fi_getpeer() call into the
tcp provider.

Signed-off-by: Sylvain Didelot <sdidelot@ddn.com>

### CODE REVIEW

File: prov/tcp/src/tcpx.h [hunk 1]
```diff
@@ -155,6 +155,7 @@ struct tcpx_ep {
 	fastlock_t		lock;
 	tcpx_ep_progress_func_t progress_func;
 	bool			send_ready_monitor;
+	struct sockaddr_in	*dest_addr;
```
Comment (shefty) [inline prov/tcp/src/tcpx.h]: We need to support ipv6.  But do we really need to store this with the EP, versus just retrieving it from the socket when needed?

File: prov/tcp/src/tcpx_ep.c [hunk 2]
```diff
@@ -483,11 +490,25 @@ static int tcpx_ep_getname(fid_t fid, void *addr, size_t *addrlen)
...
+	struct tcpx_ep *tcpx_ep;
+	size_t len;
+
+	tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+	if (!tcpx_ep->dest_addr)
+		return -FI_EINVAL;
+	len = MIN(*addrlen, sizeof(struct sockaddr_in));
+	memcpy(addr, tcpx_ep->dest_addr, len);
+	*addrlen = sizeof(struct sockaddr_in);
+	return (len == sizeof(struct sockaddr_in)) ? 0 : -FI_ETOOSMALL;
```
Comment (shefty) [inline prov/tcp/src/tcpx_ep.c]: Why not just call getpeername() here and avoid the extra complexity of storing it?  We can pass the caller's addr/addrlen values directly to the call.

Also, IPv6 support is a requirement at this point.

**Comment by sydidelot** [comment]
Thanks for the review, @shefty.
I agree, there is no need for such a complexity. I wrote my initial revision of the patch based on the getpeer() implementation in the sockets provider, which was a terribly bad idea.
I've uploaded a new revision, which should addresses all your comments. I've only tested the code with Linux. For Windows, I wrote the function `ofi_getpeername()` based on the existing `ofi_getsockname()` function, but I couldn't test it.

**Comment by shefty** [comment]
Thanks - I'm not sure that windows needs the extra complexity, but you copying what was done for getsockname() is reasonable.

**Comment by sydidelot** [comment]
Thanks!

**Pull request merged**